### PR TITLE
pk-transaction: Improvements to Finished signals

### PR DESCRIPTION
**Pull request by pwithnall**
See the commit messages for details.

### CODE REVIEW

**Comment by pwithnall** [comment]
Marking as a draft due to the open question in the third commit. This is ready for review already, though, and hopefully that should help answer the open question. I’ll update the PR once we’re happy with the approach.

**Comment by Conan-Kudo** [comment]
Posting the question from the commit:

> It feels like transactions should always emit `Finished` before being
> removed from the bus, either signalling success or containing error
> details. Emitting a signal like this allows client code to track
> transactions properly.
> 
> A lot of the code paths in `PkTransaction` ensure that `Finished` is
> emitted. However, there is at least one code path where it’s not, and
> the assertion in this commit fails:
> ```
> 4  0x00007ffff77408ae in g_assertion_message_expr () at /lib64/libglib-2.0.so.0
> 5  0x000000000044f9e7 in pk_transaction_dispose (object=0x62500010af80)
>    at ../../source/PackageKit/src/pk-transaction.c:5512
> 6  0x00007ffff7eeb8d4 in g_object_unref () at /lib64/libgobject-2.0.so.0
> 7  0x00000000004acb4b in pk_scheduler_item_free (item=0x607000311210)
>    at ../../source/PackageKit/src/pk-scheduler.c:201
> 8  0x00000000004ad140 in pk_scheduler_remove_internal
>    (scheduler=0x61d000004120, item=0x607000311210)
>    at ../../source/PackageKit/src/pk-scheduler.c:233
> 9  0x00000000004b0083 in pk_scheduler_no_commit_cb (user_data=0x607000311210)
>    at ../../source/PackageKit/src/pk-scheduler.c:590
> ```
> 
> That’s the code path to handle a transaction which is created, but which
> then isn’t used by the client before a timeout triggers.
> 
> The documentation isn’t clear on when `Finished` should be emitted.
> There’s also the `Destroy` signal which *is* always emitted before the
> transaction is removed from the bus.
> 
> When should `Finished` be used, and when should `Destroy` be used?

It seems like it would make sense that `Finished` would be emitted with successful interactions, and otherwise it would not? I guess that's why this doesn't emit one.

**Comment by pwithnall** [comment]
> It seems like it would make sense that `Finished` would be emitted with successful interactions, and otherwise it would not? I guess that's why this doesn't emit one.

The `Finished` signal has a `PkExitEnum` argument (which can take values like `PK_EXIT_ENUM_FAILED`, `PK_EXIT_ENUM_KILLED`, `PK_EXIT_ENUM_CANCELLED`, etc.), and it’s explicitly emitted on numerous error code paths in `pk-transaction.c` (see calls to `pk_transaction_finished_emit()`).

**Comment by pwithnall** [comment]
OK, updated the commit message to reflect that. Ready for review/merge from my point of view.

**Comment by ximion** [comment]
> Then I don't see why we wouldn't want to emit a `Finished` signal. 🤷🏾

@hughsie Do you remember whether there was any original reasoning here? I also believe emitting `Finished` and time a transaction is "done" is the right call, but the fact that it wasn't for so long is odd to me (unless that was just an oversight).

In any case, I'm pretty sure this change will not break any existing code :-)

**Comment by hughsie** [comment]
I believe it was an oversight.

**Review by Conan-Kudo — APPROVED** [review]

**Review by ximion — APPROVED** [review]
I've played with it a bit and couldn't find any issues - clients work as intended.
Also, the reasoning for this patch is sound :-)

Thank you for the patch!

**Comment by aleasto** [comment]
This assert triggers for transactions that return immediatly with error, as they do not emit finish _finished_.
Should they also emit _finished_, or should the assert be changed to ignore transactions with state `PK_TRANSACTION_STATE_ERROR`?

**Comment by pwithnall** [comment]
> Should they also emit _finished_, or should the assert be changed to ignore transactions with state `PK_TRANSACTION_STATE_ERROR`?

Read the comments higher up on this PR: we decided that transactions should always emit `Finished` and there was no benefit to making it more complex than that.

So that code should be changed to emit `Finished`.

File: lib/packagekit-glib2/pk-client.c [hunk 1]
```diff
@@ -1582,8 +1587,16 @@ pk_client_signal_cb (GDBusProxy *proxy,
...
 	}
-	if (g_strcmp0 (signal_name, "Destroy") == 0)
+	if (g_strcmp0 (signal_name, "Destroy") == 0) {
+		g_autoptr(GError) local_error = NULL;
+
+		if (state->waiting_for_finished)
+			local_error = g_error_new_literal (PK_CLIENT_ERROR, PK_CLIENT_ERROR_FAILED,
+							   "PackageKit transaction disappeared");
+
+		pk_client_state_finish (state, local_error);
```
[the remark this answers is not in the document]
Comment (mcrha) [reply lib/packagekit-glib2/pk-client.c:1597]: I opened https://github.com/PackageKit/PackageKit/pull/774 to fix it

**Pull request merged**